feat(trip-explorer): speed for station pairs

rudiejd opened the pull request:
## Motivation

Closes https://github.com/transitmatters/t-performance-dash/issues/803

<!-- Why are you making this change, what problem does it solve? Include links to relevant issues -->

## Changes
- Add speed graph on trip explorer

<!-- What does this change exactly? Include relevant screenshots, videos, links -->
![image](https://github.com/transitmatters/t-performance-dash/assets/46619169/ccc5e3e9-9e2c-4907-b168-2638fedba5f2)




## Testing Instructions
- Navigate to trip explorer page for all lines
- Verify that the new speed graph is populated correctly

<!-- How can the reviewer confirm these changes do what you say they do? -->

rudiejd commented:
this is really speed for stop pairs if we're being technical 🤓 

I will try to fix the nomenclature to consistently call them stops

rudiejd commented on common/constants/stations.json:
```diff
@@ -641,8 +899,12 @@
         "station": "place-mvbcl",
         "order": 8,
         "stops": {
```
I think this auto-formatted because I opened it 😮‍💨 

I wanted to add distances here, but I figured it would be either to make another file. Let me know your thoughts

devinmatte replied:
The other file seems ideal given how large it is 👍

rudiejd commented on server/chalicelib/speed.py:
```diff
@@ -80,3 +82,6 @@ def is_invalid_range(start_date, end_date, max_delta):
     end_datetime = datetime.strptime(end_date, DATE_FORMAT_BACKEND)
     return start_datetime + timedelta(days=max_delta) < end_datetime
+
+if __name__ == 'main':
```
[ ] remove

rudiejd commented on server/scripts/update_stations.py:
```diff
@@ -74,5 +74,112 @@
+        seen_stops = set()
+        
+        # assume each stop starts with one destination 
+        first_dest, first_dist = next(iter(stop_distances[target_stop_id].items()))
```
you could probably make this better by doing a seperate DFS for each route 🤷

rudiejd commented on modules/landing/utils.ts:
```diff
@@ -45,6 +47,24 @@ export const convertToSpeedDataset = (data: { [key in Line]?: DeliveredTripMetri
+  data: SingleDayDataPoint[]
+) => {
+  const intervalDistance = getDistancesForStopId(fromStationId, toStationId);
+  console.log(`${fromStationId}, ${toStationId}, ${intervalDistance}`);
```
- [ ] remove

devinmatte reviewed (commented):
This is awesome! A few initial thoughts in no particular order:

We should display both of the two station pairs since this is between two
<img width="1201" alt="Screenshot 2023-12-17 at 7 38 49 PM" src="https://github.com/transitmatters/t-performance-dash/assets/9310513/63711f4f-51a0-4f9e-a676-b6102a7213ea">
<img width="185" alt="Screenshot 2023-12-17 at 7 38 52 PM" src="https://github.com/transitmatters/t-performance-dash/assets/9310513/6829ebd0-0f10-4371-982a-e2112e95f68e">

We should add the same aggregate numbers below the chart like we do in travel times
<img width="1175" alt="Screenshot 2023-12-17 at 7 38 40 PM" src="https://github.com/transitmatters/t-performance-dash/assets/9310513/d32fb73f-d612-42b6-aec5-ff1d8cc2bb0e">

Personally I think speed should go between headways and dwells in the chart order

I was able to get the page to bug out with `http://localhost:3000/red/trips/single/?from=70073&to=70093&date=2023-12-17` so there's probably some missing stops in the dataset?

Overall, this is very cool, just needs some polish. We can try and get this into beta before the next meeting to get people in labs to poke around with it

rudiejd commented:
> This is awesome! A few initial thoughts in no particular order:
> 
> We should display both of the two station pairs since this is between two <img alt="Screenshot 2023-12-17 at 7 38 49 PM" width="1201" src="https://private-user-images.githubusercontent.com/9310513/291116923-63711f4f-51a0-4f9e-a676-b6102a7213ea.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI4NjEyOTAsIm5iZiI6MTcwMjg2MDk5MCwicGF0aCI6Ii85MzEwNTEzLzI5MTExNjkyMy02MzcxMWY0Zi01MWEwLTRmOWUtYTY3Ni1iNjEwMmE3MjEzZWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIxOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMThUMDA1NjMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTZhZmU2M2QzNjAzYjM2YTBjYzkxZjA1Y2M4MWEyMDZmNzY2NmMwNTM1YjAyNmEyODk4NjlkNzI1Yjc5ZGYxZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.tEn7IPUn0Vodcn7r7jZ3VrWc0jXKwOWpUKjfoFHS3cg"> <img alt="Screenshot 2023-12-17 at 7 38 52 PM" width="185" src="https://private-user-images.githubusercontent.com/9310513/291116924-6829ebd0-0f10-4371-982a-e2112e95f68e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI4NjEyOTAsIm5iZiI6MTcwMjg2MDk5MCwicGF0aCI6Ii85MzEwNTEzLzI5MTExNjkyNC02ODI5ZWJkMC0wZjEwLTQzNzEtOTgyYS1lMjExMmU5NWY2OGUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIxOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMThUMDA1NjMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzM3ZWMwOTNhMGE0MzZiY2ZjZmNkYjNlODk5ODUwMWVhZTU1ZTE5MzY2ODBjZmQzMmU3NTlkODBjYTc0ZWJkMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.7Iip-kS2nUbVWW-8efx3B68Qff4G2q7G8tlYTIvnSAU">
> 
> We should add the same aggregate numbers below the chart like we do in travel times <img alt="Screenshot 2023-12-17 at 7 38 40 PM" width="1175" src="https://private-user-images.githubusercontent.com/9310513/291116922-d32fb73f-d612-42b6-aec5-ff1d8cc2bb0e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI4NjEyOTAsIm5iZiI6MTcwMjg2MDk5MCwicGF0aCI6Ii85MzEwNTEzLzI5MTExNjkyMi1kMzJmYjczZi1kNjEyLTQyYjYtYWVjNS1mZjFkOGNjMmJiMGUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIxOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMThUMDA1NjMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDUyZTY4NTIxZjY3ZDkzY2I5OWUwM2UyMTEzNTI3YzY5MjlmMTM4NGNmZWI0YmNmYWZlMjNiOTViNDVmMmVhOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.4akkZip-oM9zWT2bwdSts58qQsoQm4c1AP2hoO7723g">
> 
> Personally I think speed should go between headways and dwells in the chart order
> 
> I was able to get the page to bug out with `http://localhost:3000/red/trips/single/?from=70073&to=70093&date=2023-12-17` so there's probably some missing stops in the dataset?
> 
> Overall, this is very cool, just needs some polish. We can try and get this into beta before the next meeting to get people in labs to poke around with it

thanks for taking a look!! yep, this is definitely still a WIP - just threw it together yesterday/today. Those are all good. There were a couple other things I wanted to take a look at too, let me know your thoughts: 
- some of the speeds seem a bit suspect, for example average of 20mph some hours for davis to quincy center. I wonder if I should be including the dwells in addition to `travel_time_sec`
- there is no monthly/weekly aggregated version of this chart
- there is no benchmark. I could calculate a benchmark using the benchmark travel time. 
- the distance calculations from traversing that list are right, but unfortunately the dataset includes some bizarre distances - for example, JFK / UMass (Braintree) => Ashmont (Ashmont) was listed as 20 miles. I haven't debugged this super deeply yet

devinmatte commented:
> there is no monthly/weekly aggregated version of this chart

This certainly isn't necessary for a first pass. We can start with a single day chart only

> there is no benchmark. I could calculate a benchmark using the benchmark travel time.

there's no benchmark for dwells, so it's not critical

> some of the speeds seem a bit suspect, for example average of 20mph some hours for davis to quincy center. I wonder if I should be including the dwells in addition to travel_time_sec

Yeah maybe @PatrickCleary or @idreyn could help you track down the accuracy of these since they worked on the line-wide speed numbers. I can also poke around after the holiday

rudiejd commented:
> I was able to get the page to bug out with http://localhost:3000/red/trips/single/?from=70073&to=70093&date=2023-12-17 so there's probably some missing stops in the dataset?

I poked around a bit on this. It seems that `70073` is Charles MGH on the Alewife - Braintree route and `70093` is Ashmont on the Alewife - Ashmont route. By this dataset, that is unreachable since you can't access `70093` on the same route. I originally used `stop_id` for the distance calculation since it was readily available on the FE, but it seems like it might be better to use `station_id` for the distances based on this quirk 😅

rudiejd commented:
I changed this to use stations instead of stops, fixing the Charles => Ashmont issue, fixed the distance calculation by only traversing in one direction at a time, added a legend, and updated the screenshots. 

I also added a benchmark speed calculation. I can show it on the graph if that's something we want, but I think it might be a little confusing considering it'll be the inverse of the travel time benchmark

devinmatte commented:
Damn seeing full speeds for tracks that are repaired is wild
<img width="1204" alt="Screenshot 2023-12-20 at 8 25 18 PM" src="https://github.com/transitmatters/t-performance-dash/assets/9310513/20c1c5ee-d0e4-4041-a86e-5c3839272353">

I do think this will be more interesting in aggregate (multi-day) datasets to see improvement/decline than it will be for most single day views (since it really is just another way to describe travel times) but this is quite interesting to see when a track is fixed.

rudiejd commented:
![image](https://github.com/transitmatters/t-performance-dash/assets/46619169/d65d4a76-dfbc-4f09-930c-b4a1250063f2)
Added aggregate data.

One small issue is that some of the green line station distances e.g. Ball Square => Back of the Hill are still undefined, and the aggregate charts don't accept undefined data, so you end up with these weird graphs of 0 speed. I'll

PatrickCleary commented:
The green line station distances we use are just measured on Google Maps. So if you're interested in replacing some of those null values, you could add them that way!

rudiejd commented:
@PatrickCleary  I ended up figuring out the problem with the data - I was relying on the  `route-id` field being the line name, but it was not `Green` for all of the green, but rather `Green-A`, `Green-B`, etc. but thank you! Green line looks to be fixed now
![image](https://github.com/transitmatters/t-performance-dash/assets/46619169/0063cfe1-7f20-417a-8841-56cc4989fb34)

I refactored the script for calculating station distances and moved it out to its own file, because the logic was getting pretty complicated.

devinmatte commented on modules/tripexplorer/SubwayTripGraphs.tsx:
```diff
@@ -62,6 +64,20 @@ export const SubwayTripGraphs: React.FC<SubwayTripGraphsProps> = ({
+          <WidgetDiv>
+            <WidgetTitle
+              title="Speeds"
+              subtitle="Speed between stops"
```
We should say 
```suggestion
              subtitle="Average speed between stops"
```
since there are several speed limits between a pair

devinmatte commented on modules/tripexplorer/SubwayTripGraphs.tsx:
```diff
@@ -130,6 +146,21 @@ export const SubwayTripGraphs: React.FC<SubwayTripGraphsProps> = ({
+          <WidgetDiv>
+            <WidgetTitle
+              title="Speeds"
+              subtitle="Speed between stops"
```
Same here 
```suggestion
              subtitle="Average speed between stops"
```

devinmatte commented on server/pyproject.toml:
```diff
@@ -32,3 +32,10 @@ target-version = ['py311']
 requires = ["poetry-core>=1.6.0"]
 build-backend = "poetry.core.masonry.api"
+
+[tool.pyright]
+reportMissingImports = true
+reportMissingTypeStubs = false
+useLibraryCodeForTypes = false
```
I'm guessing this is from local environment, we probably don't want this committed

rudiejd replied:
this is actually  setting up the language server I use for neovim, [pyright](https://github.com/microsoft/pyright) to not scream at me because we don't use type stubs - it is in strict type-checking mode by default. it shoudn't affect anything other than pyright, and the settings should be universal (you can think of it like `.editorconfig`) but I can remove if it's confusing

devinmatte replied:
Okay maybe @nathan-weinberg has an opinion about us keeping this in or removing then

nathan-weinberg replied:
If the additional config indeed only affects `pyright` and won't have any other effects (i.e. no change to the project) I don't see the harm in leaving this in - don't think we have any other devs using this so we can always adjust if that ever changes

devinmatte commented on modules/tripexplorer/SubwayTripGraphs.tsx:
```diff
@@ -62,6 +64,20 @@ export const SubwayTripGraphs: React.FC<SubwayTripGraphsProps> = ({
               toStation={toStation}
             />
           </WidgetDiv>
+          <WidgetDiv>
```
Since this is basically an inverse of travel times, we shouldn't show it back to back with travel times. We should move it below Headways at a minimum. Probably after dwells imo

devinmatte commented on modules/tripexplorer/SubwayTripGraphs.tsx:
```diff
@@ -130,6 +146,21 @@ export const SubwayTripGraphs: React.FC<SubwayTripGraphsProps> = ({
             />
           </WidgetDiv>
 
+          <WidgetDiv>
```
Same here, below dwells ideally, definitely below headways

devinmatte reviewed (commented):
Some initial comments

austinjpaul commented:
Mentioned in the meeting, just writing here for not forgetting - I think my preference would be to add speed as a toggle for the travel-times chart, since it is the same data displayed two different ways, just scaled by a factor. That way we don't end up with too many charts on page.

rudiejd commented:
@austinjpaul totally agree! I'll rework it as a toggle on the travel times graph

rudiejd commented:
I reworked this to make speed a toggle on the travel times graph. I currently have it set as the first option for travel time, because I think it's the easier to read given our users might not know what the travel times for a given station pair should be

rudiejd commented:
![image](https://github.com/transitmatters/t-performance-dash/assets/46619169/406bc532-1d7a-4d78-ab3c-b7a8de9e9938)

rudiejd commented:
I switched the graph around to make travel times the default since that's what users are more used to following comments from @austinjpaul in slack. This should be ready to go now

devinmatte commented on server/poetry.lock:
I'm not sure why the lock file is getting updated here, especially with py12 binaries

rudiejd replied:
no clue either. It's adding a PyYaml binary for musl, but I use glibc on my machine in WSL. I can remove this

devinmatte commented on server/scripts/update_stations.py:
```diff
@@ -74,5 +73,5 @@
 
 
-with open("common/constants/stations.json", "w") as f:
+with open("common/constants/stations_distances.json", "w") as f:
```
This doesn't seem necessary

rudiejd replied:
good catch, that actually would have broken this script 💀

devinmatte reviewed (commented):
Looking good, two small things but overall, basically ready to approve
